#!/usr/bin/env python
PACKAGE = "sentry_chassis_controller"
from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# Wheel PID with Low-pass filter
# 低通滤波器时间常数 (tau): 值越大滤波越强，响应越慢；值越小响应越快
# 建议范围: 0.001 ~ 0.1 秒，0 表示禁用滤波
for w in ["wheel_fl", "wheel_fr", "wheel_rl", "wheel_rr"]:
    gen.add(w + "_p", double_t, 0, w + " P", 40, 0.0, 100.0)
    gen.add(w + "_i", double_t, 0, w + " I", 0.005, 0.0, 100.0)
    gen.add(w + "_d", double_t, 0, w + " D", 0.0015, 0.0, 100.0)
    gen.add(w + "_i_clamp", double_t, 0, w + " I clamp", 1, 0.0, 100.0)
    gen.add(w + "_output_min", double_t, 0, w + " output min", -10.0, -1000.0, 0.0)
    gen.add(w + "_output_max", double_t, 0, w + " output max", 10.0, 0.0, 1000.0)
    gen.add(w + "_lpf_tau", double_t, 0, w + " LPF time constant (s)", 0.02, 0.0, 0.5)

for p in ["pivot_fl", "pivot_fr", "pivot_rl", "pivot_rr"]:
    gen.add(p + "_p", double_t, 0, p + " P", 1.3, 0.0, 100.0)
    gen.add(p + "_i", double_t, 0, p + " I", 0.03, 0.0, 100.0)
    gen.add(p + "_d", double_t, 0, p + " D", 0.035, 0.0, 100.0)
    gen.add(p + "_i_clamp", double_t, 0, p + " I clamp", 0.2, 0.0, 100.0)
    gen.add(p + "_output_min", double_t, 0, p + " output min", -5.0, -1000.0, 0.0)
    gen.add(p + "_output_max", double_t, 0, p + " output max", 5.0, 0.0, 1000.0)
    gen.add(p + "_antiwindup", double_t, 0, p + " antiwindup", 0.0, 0.0, 100.0)

# Power Limiter Parameters
gen.add("effort_coeff", double_t, 0, "Power limiter effort coefficient", 18.0, 0.0, 200.0)
gen.add("velocity_coeff", double_t, 0, "Power limiter velocity coefficient", 0.0048, 0.0, 1.0)
gen.add("power_limit", double_t, 0, "Power limit (W)", 480.0, 0.0, 1000.0)
gen.add("power_offset", double_t, 0, "Power offset (W)", 0.0, 0.0, 500.0)

exit(gen.generate(PACKAGE, "sentry_chassis_controller", "WheelPid"))
